### PR TITLE
Consolidate abstractions and core types into go-libp2p-core

### PR DESCRIPTION
**Pull request by raulk**
This PR is part of a batch across go-libp2p repos to consolidate abstractions, interfaces and core types into [libp2p/go-libp2p-core](https://github.com/libp2p/go-libp2p-core). See branch [`consolidate-skeleton`](https://github.com/libp2p/go-libp2p-core/tree/consolidate-skeleton) there.

Work is being performed in the `consolidate-skeleton` branch across repos.

In this PR, we:
1. migrate core types out of this repo into go-libp2p-core.
2. ensure backwards compatibility by aliasing all items that were publicly exported.
3. introduce a gomod dependency to go-libp2p-core.

---

Merge strategy: squash and rebase.

### CODE REVIEW

File: interface.go [hunk 1]
```diff
@@ -1,26 +1,12 @@
 package discovery
 
-import (
-	"context"
-	"time"
+import moved "github.com/libp2p/go-libp2p-core/discovery"
 
-	pstore "github.com/libp2p/go-libp2p-peerstore"
-)
+// Deprecated: use skel.Advertiser instead.
```
Comment (vyzo) [inline interface.go]: what the hell is `skel`?

File: options.go [hunk 2]
```diff
@@ -2,40 +2,27 @@ package discovery
 
 import (
 	"time"
-)
 
-// Option is a single discovery option
-type Option func(opts *Options) error
+	moved "github.com/libp2p/go-libp2p-core/discovery"
```
Comment (vyzo) [inline options.go]: I would call this something else in here, `moved` is ugly and we use these types.
Perhaps `iface`?

File: options.go [hunk 3]
```diff
@@ -2,40 +2,27 @@ package discovery
...
-	}
-	return nil
-}
+// Deprecated: use skel.DiscoveryOpts instead.
+type Options = moved.DiscoveryOpts
 
 // TTL is an option that provides a hint for the duration of an advertisement
-func TTL(ttl time.Duration) Option {
-	return func(opts *Options) error {
+func TTL(ttl time.Duration) moved.DiscoveryOpt {
```
Comment (vyzo) [inline options.go]: shouldn't the options move too?

File: options.go (hunk 2 above)
Comment (raulk) [reply options.go]: `iface` won't work as the target packages doesn't only contain interfaces. I think `moved` is fine as it clearly conveys intent. Alternatively `target`, but that's less specific.

File: options.go (hunk 3 above)
Comment (raulk) [reply options.go]: SGTM; they are not implementation-specific.

What about the util methods under `util.go`? They are implementation-agnostic.

Comment (vyzo) [reply options.go]: I think the whole package can be moved actually. It's all utility functions with no dependency on concrete implementations.

File: options.go (hunk 2 above)
Comment (vyzo) [reply options.go]: Just call it `core`!

File: options.go (hunk 3 above)
Comment (vyzo) [reply options.go]: Actually let's keep `routing` here -- we will eventually add rendezvous too.
But the utilities can certainly move.

File: options.go (hunk 2 above)
Comment (raulk) [reply options.go]: Thinking of it, `core` is not half-bad, as it conveys where the type lives now as well.

File: interface.go (hunk 1 above)
Comment (raulk) [reply interface.go]: Used to mean "skeleton", this is outdated now.

File: options.go (hunk 2 above)
Comment (raulk) [reply options.go]: Ok, you won, I've changed it everywhere to `core`.

File: options.go (hunk 3 above)
Comment (raulk) [reply options.go]: Hm, but routing doesn't deal with implementation-specific details, it just deals with abstractions. In fact, go-libp2p-rendezvous should just satisfy the Discovery interface. I think this repo should go away.

**Pull request merged**